tests/output.rs

```diff
@@ -0,0 +1,174 @@
+}
+
+fn predicate_from_expected_match(expected_match: &ExpectedMatch) -> BoxPredicate<str> {
+    BoxPredicate::new(predicate::str::contains(
+        expected_match.expected_output_text(),
+    ))
+}
+
+fn assert_expected_matches<TArg: AsRef<OsStr>>(
+    args: impl IntoIterator<Item = TArg>,
+    current_dir: impl AsRef<Path>,
+    expected_matches: &[ExpectedMatch],
+) {
+    Command::cargo_bin("tree-sitter-grep")
+        .unwrap()
+        .args(args)
+        .current_dir(current_dir)
+        .assert()
+        .success()
+        .stdout(predicate_from_expected_matches(expected_matches));
+}
+
+struct FixtureQueryExpectedMatches {
+    pub fixture_dir_name: String,
+    pub language: String,
+    pub query_source: String,
+    pub expected_matches: Vec<ExpectedMatch>,
+}
+
+impl FixtureQueryExpectedMatches {
+    pub fn assert_expected_matches(&self) {
+        assert_expected_matches(
+            [
+                "--query-source",
+                &self.query_source,
+                "--language",
+                &self.language,
+            ],
+            get_fixture_dir_path_from_name(&self.fixture_dir_name),
+            &self.expected_matches,
+        );
+    }
+}
+
+fn get_fixture_dir_path_from_name(fixture_dir_name: &str) -> PathBuf {
+    // per https://andrewra.dev/2019/03/01/testing-in-rust-temporary-files/
+    let root_dir = env::var("CARGO_MANIFEST_DIR").unwrap();
+    let mut path: PathBuf = root_dir.into();
+    path.push("tests/fixtures");
+    path.push(fixture_dir_name);
+    path
+}
+
+const FUNCTION_ITEM_QUERY_SOURCE: &'static str = "(function_item) @function_item";
+
+static RUST_PROJECT_FUNCTION_ITEM_EXPECTED_MATCHES: Lazy<FixtureQueryExpectedMatches> =
+    Lazy::new(|| FixtureQueryExpectedMatches {
+        fixture_dir_name: "rust_project".into(),
+        query_source: FUNCTION_ITEM_QUERY_SOURCE.to_owned(),
+        language: "rust".to_owned(),
+        expected_matches: vec![
+            ExpectedMatch {
+                relative_file_path: "./src/helpers.rs".to_owned(),
+                lines: vec!["pub fn helper() {}".to_owned()],
+                range: ((0, 0), (0, 18)).into(),
+            },
+            ExpectedMatch {
+                relative_file_path: "./src/lib.rs".to_owned(),
+                lines: vec![
+                    "pub fn add(left: usize, right: usize) -> usize {".to_owned(),
+                    "    left + right".to_owned(),
+                    "}".to_owned(),
+                ],
+                range: ((2, 0), (5, 0)).into(),
+            },
+            ExpectedMatch {
+                relative_file_path: "./src/lib.rs".to_owned(),
+                lines: vec![
+                    "    fn it_works() {".to_owned(),
+                    "        let result = add(2, 2);".to_owned(),
+                    "        assert_eq!(result, 4);".to_owned(),
+                    "    }".to_owned(),
+                ],
+                range: ((11, 0), (14, 4)).into(),
+            },
```

Comment:
This reads nicely, but if you wanted to get really crazy/cute with it, you could write a little test parser a-la-rust-analyzer tests? I could imagine writing tests like this?

```rs
#[test]
fn test_something() {
assert_output!(r#"
$ tree-sitter-grep --some-option
.src/helpers.rs:10:   pub fn helper() {}
...
"#);
```

Where you write `$ the command to actually execute` followed by the expected output as string.

(I don't know that it's actually a good idea, just something that occurred to me.)

Reply:
Snapshot tests also seems appealing for something like this.

Reply:
Ah yeah, I'd also been assuming without looking into it that the ordering was not deterministic. I think the snapshotter could sort the output, though, and then it'd be back to deterministic?

> so does it not do something similar to what we're doing and then also do the matching in parallel (without then collecting/sorting the aggregated matches)?

Huh, yeah I'd have assumed the matching was also in parallel.